spellbook/models/sudoswap/ethereum/sudoswap_ethereum_events.sql

```diff
@@ -0,0 +1,240 @@
+            , _fee as initialfee
+            , _assetRecipient as asset_recip
+            , output_pair as pair_address
+            , call_block_time as block_time
+        FROM {{ source('sudo_amm_ethereum','LSSVMPairFactory_call_createPairETH') }}
+        WHERE call_success
+    ),
+
+    owner_fee_update as (
+        SELECT 
+            *
+        FROM {{ source('sudo_amm_ethereum','LSSVMPair_general_evt_FeeUpdate') }}
+    ),
+
+    protocol_fee_update as (
+        SELECT
+            *
+        FROM {{ source('sudo_amm_ethereum','LSSVMPairFactory_evt_ProtocolFeeMultiplierUpdate') }}
+    ),
+
+    eth_traces as (
+        SELECT 
+            * 
+        FROM {{ source('ethereum', 'traces') }}
+    ),
+
+    eth_transactions as (
+        SELECT 
+            *
+        FROM {{ source('ethereum','transactions') }}
+    ),
+
+    tokens_ethereum_nft as (
+        SELECT 
+            *
+        FROM {{ ref('tokens_ethereum_nft') }}
+    ),
+
+    nft_ethereum_aggregators as (
+        SELECT 
+            *
+        FROM {{ ref('nft_ethereum_aggregators') }}
+    ),
+
+    prices_usd_eth as (
+        SELECT 
+            *
+        FROM {{ source('prices','usd') }}
+        WHERE symbol = 'WETH'
+    ),
+
+--logic CTEs
+    swaps_w_fees as (
+        SELECT 
+            *
+        FROM (
+            SELECT
+                call_tx_hash
+                , call_block_time
+                , call_block_number
+                , contract_address as pair_address
+                , call_trace_address
+                , ownerfee 
+                , protocolfee 
+                , trade_category
+                , nftcontractaddress
+                , asset_recip
+                , call_from
+                , row_number() OVER (partition by call_tx_hash, contract_address, call_trace_address order by fee_update_time desc, protocolfee_updatetime desc) as ordering
+            FROM (
+                SELECT 
+                    swaps.*
+                    , COALESCE(fu.newfee, pc.initialfee)/1e18 as ownerfee --most recent ownerfee, depends on bonding curve to implement it correctly. See explanation in fee table schema.
+                    , COALESCE(fu.evt_block_time, pc.block_time) as fee_update_time
+                    , pfu.newMultiplier/1e18 as protocolfee --most recent protocolfee, depends on bonding curve to implement it correctly. See explanation in fee table schema.
+                    , pfu.evt_block_time as protocolfee_updatetime
+                    , pc.nftcontractaddress
+                    , pc.asset_recip
+                FROM swaps
+                JOIN pairs_created pc ON pc.pair_address = contract_address --remember swaps from other NFT addresses won't appear!
+                LEFT JOIN owner_fee_update fu ON swaps.call_block_time >= fu.evt_block_time AND swaps.contract_address = fu.contract_address
+                LEFT JOIN protocol_fee_update pfu ON swaps.call_block_time >= pfu.evt_block_time
```

Comment:
without looking closer, is block time the only applicable join condition? should there be more?

Reply:
nope, the protocol fee is updated across all pair contracts (checks against the factory)